### PR TITLE
introduce analyzer phase

### LINKED ISSUE #132: Logical error handling: Fail quickly when SELECT/CONSTRUCT vars do not exist in WHERE clause
Expected behavior: Query should fail and exit before execution begins if variable is referenced in SELECT or CONSTRUCT that does not exist in WHERE clauses.

Current behavior: Spark execution begins and only fails when execution of WHERE is complete.

Example 
SELECT ?species_node WHERE {
        <http://purl.obolibrary.org/obo/CLO_0037232> <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?derived_node .
}

image

### PR DESCRIPTION
**Pull request by pepegar**
In this phase we'll catch logical errors such as the one described in #132.

Closes #132

### CODE REVIEW

File: modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/Analyzer.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+    * Execute all rules in [[Analyzer.rules]] and accumulate errors
+    * that they may throw.
+    *
+    * In case no errors are returned, the
+    *
+    * @return
+    */
+  def analyze[T: Basis[DAG, *]]: Phase[T, T] =
```
Comment (patofp) [inline modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/Analyzer.scala:32]: Great idea using Validated this way we can stack validation errors at once!

File: modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/package.scala [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.gsk.kg.engine
+
+import cats.data.ValidatedNec
+
+package object analyzer {
+
+  type Rule[T] = T => ValidatedNec[String, String]
```
Comment (patofp) [inline modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/package.scala:7]: Nice!

**Review by patofp — APPROVED** [review]
LGTM!

File: modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/Analyzer.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+      x match {
+        case Invalid(e) => M.lift[Result, DataFrame, T](EngineError.AnalyzerError(e).asLeft)
+        case Valid(e) => t.pure[M]
+      }
+    }
+
+  def findUnboundVariables[T](implicit T: Basis[DAG, T]): Rule[T] = { t =>
+    val declaredVariables: Set[VARIABLE] = t.collect[List[List[VARIABLE]], List[VARIABLE]] {
+        case BGP(triples) =>
```
Comment (shihch) [inline modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/Analyzer.scala:44]: Besides BGP, Extend, FilterLeftJoin, Filter, and nested Project could introduce variables. Do we have to consider all of them?

Comment (pepegar) [reply modules/engine/src/main/scala/com/gsk/kg/engine/analyzer/Analyzer.scala]: Thanks @shihch good point.

I've changed the approach for doing the analysis to a full bottom-up traverse checking variables declared level by level

**Review by patofp — APPROVED** [review]
LGTM!

**Pull request merged**